ponzu add <addon URI>

olliephillips opened the pull request:
Reference issue #85. As discussed, this adds a `ponzu add` command to the CLI. It fetches the addon from its URI by wrapping `go get` then copies it to the projects local `./addons` directory. 

There is no switch for automatically installing. Based on the issue discussion that feature needs more thought, so has been omitted.

I've updated the README to include documentation on the command. 

It's been tested as per the approach in the README.

nilslice commented on cmd/ponzu/add.go:
```diff
@@ -0,0 +1,177 @@
+
+	// Go get
+	cmdOptions = append(cmdOptions, "get", addonPath)
+	get := exec.Command(gocmd, cmdOptions...)
```
good catch to re-use the gocmd variable.. will be very useful when working with future go version betas / release candidates. thank you for being mindful about that!

nilslice commented on cmd/ponzu/add.go:
```diff
@@ -0,0 +1,177 @@
+	// copy to ./addons folder
+	// GOPATH can be a list delimited by ":" on Linux or ";" on Windows
+	// `go get` uses the first, this should parse out the first whatever the OS
+	gopath := resolveGOPATH()
```
if you don't mind, before we merge this in, could you take a look at #82 and see how it would impact GOPATH resolution in the way you have it? 

#82 hasn't been fixed, but the code is pretty much written in the issue... if you felt inclined to add that in I would be grateful.

olliephillips replied:
Certainly. I didn't know anything about that. I only upgraded to 1.8 to run Ponzu :)

nilslice replied:
Great -- thank you very much!

nilslice commented on cmd/ponzu/usage.go:
```diff
@@ -162,6 +162,19 @@ var usageVersion = `
 `
 
+var usageAdd = `
+[--cli] add, a <addon URI>
```
is `[--cli]` actually used in the add command? if not, lets clear that so it's not confusing. then in place of <addon URI>, let's say <import-path>

olliephillips replied:
It's not in the end PR, I spotted in late and removed it in [this commit] (https://github.com/ponzu-cms/ponzu/pull/87/commits/5fcca313e9b421a6ed19881bc7d8e28bb0e507bf)

nilslice replied:
All good! Thank you. I figured it was a copy/paste thing.. and I appreciate you keeping to the style of the existing commands!

olliephillips commented:
This is what I had in mind for GOPATH resolution 

```
// resolve GOPATH. In 1.8 can be default, or custom. A custom GOPATH can
// also contain multiple paths, in which case 'go get' uses the first
func getGOPATH() (string, error) {
	var gopath string
	gopath = os.Getenv("GOPATH")
	if gopath == "" {
		// not set, find the default
		usr, err := user.Current()
		if err != nil {
			return gopath, err
		}
		gopath = filepath.Join(usr.HomeDir, "go")
	} else {
		// parse out in case of multiple, retain first
		gopaths := strings.Split(gopath, ":") // Linux
		gopath = gopaths[0]
		gopaths = strings.Split(gopath, ";") // Windows
		gopath = gopaths[0]
	}
	return gopath, nil
}
```

Sirikon commented:
@olliephillips That script could have the same problem I commented you in the code.

nilslice commented:
@Sirikon - could you link me to the comment / line in the code you mention? I can't find it.

Sirikon commented:
@nilslice https://github.com/ponzu-cms/ponzu/pull/87/files/88dc9b4d26022aeda295e03690c4eb4398e40487#diff-88cf37571f8025a3c6cc4fa53928c999R62

nilslice commented:
@Sirikon - sorry, I meant the link to your comment that you left about:
> That script could have the same problem I commented you in the code.

I don't see any comment there, maybe I'm just missing it..

Sirikon commented:
@nilslice This one you mean? https://github.com/ponzu-cms/ponzu/pull/87/files#r103056524

nilslice commented:
@Sirikon - odd, this is what I see when clicking your link. No comment from you on the whole PR.. 
![screenshot from 2017-02-28 02 38 40](https://cloud.githubusercontent.com/assets/7517515/23402150/3382b7fc-fd5f-11e6-8ccb-bd7afdeb48c9.png)

Would you mind copying the comment you made and the line # you're referencing so I can understand what the issue is?

Thanks & sorry for the inconvenience.. really not sure why I can't see the comment

Sirikon commented:
For god sake Github X_X.

Sorry guys, here it is: Line 62-69 of file add.go
![image](https://cloud.githubusercontent.com/assets/1157382/23402295/3ea26960-fdab-11e6-9a7f-9f6e74985189.png)

Sirikon commented:
The rest of the script seems perfect for me :)

olliephillips commented:
Needed an OS check in end.  Hopefully, this should reliably get GOPATH or the default. At least for Windows/Linux/Mac. If agreed I'll update the PR.

```
// resolve GOPATH. In 1.8 can be default, or custom. A custom GOPATH can
// also contain multiple paths, in which case 'go get' uses the first
func getGOPATH() (string, error) {
	var gopath string
	gopath = os.Getenv("GOPATH")
	if gopath == "" {
		// not set, find the default
		usr, err := user.Current()
		if err != nil {
			return gopath, err
		}
		gopath = filepath.Join(usr.HomeDir, "go")
	} else {
		// parse out in case of multiple, retain first
		if runtime.GOOS == "windows" {
			gopaths := strings.Split(gopath, ";")
			gopath = gopaths[0]
		} else {
			gopaths := strings.Split(gopath, ":")
			gopath = gopaths[0]
		}
	}
	return gopath, nil
}
```

Sirikon commented:
Seems okay for me :)

nilslice commented:
@sirikon - thank you for adding that, sorry it was pending.. not sure how to approve those things

@olliephillips - that looks good to me, go ahead and update the PR at your convenience. 

Thanks!

olliephillips commented:
@nilslice That has been added to the PR. I've run through the ```ponzu build``` process again and seems ok to me

nilslice commented:
@olliephillips, thank you for adding this. I think it will be a really big help to make addons a major component of Ponzu systems.

@Sirikon, thank you for your continued involvement and testing! 

I really appreciate all of your help.
